iree/turbine/kernel/wave/utils.py

```diff
@@ -1055,3 +1056,220 @@ def evaluate_with_assumptions(constraints: list[Constraint], expr: IndexExpr) ->
+
+            if not arg.is_nonnegative:
+                return None
+
+        return ret
+
+    def transform_floor(expr):
+        """Simplify rational addition inside floor expr
+
+        Example:
+        floor(floor(a)/3 + 1/6) -> floor(floor(a)/3)
+        """
+        if not isinstance(expr, sympy.floor):
+            return None
+
+        expr = expr.args[0]
+        if not isinstance(expr, sympy.Add):
+            return None
+
+        c = None
+        for arg in expr.args:
+            if isinstance(arg, sympy.Rational):
+                if c is not None:
+                    return None
+
+                c = arg
+
+        if c is None:
+            return None
+
+        terms = []
+        for arg in expr.args:
+            if isinstance(arg, sympy.Rational):
+                continue
+
+            if not isinstance(arg, sympy.Mul):
+                return None
+
+            r = check_mul_rational(arg)
+            if r is None or r.p != 1:
+                return None
+
+            if r <= c:
+                return None
+
+            terms.append(arg)
+
+        return sympy.floor(sum(terms))
+
+    expr = expr.replace(lambda e: transform_mod(e) is not None, transform_mod)
+    expr = expr.replace(lambda e: transform_floor(e) is not None, transform_floor)
+    return sympy.simplify(expr)
+
+
+def check_is_mapping_contiguous(
+    mapping: IndexMapping,
+    symbolc_shape: tuple[IndexExpr, ...],
+    index: tuple[IndexExpr, ...],
+    elements_per_thread: int | IndexExpr,
+    is_read: bool,
+) -> bool:
+    """Check if mapping can be lowered to contiguous vector ops instead of gathers/scatters"""
+    elements_per_thread = subs_idxc(elements_per_thread)
+    if elements_per_thread == 1:
+        return True
+
+    # TODO: Better dyn vals analysis.
+    if mapping.num_dynamic_vals != 0:
+        return False
+
+    if is_read:
+        assert (
+            mapping.is_output_identity()
+        ), "non-identity output mapping is not supported yet"
+        index_mapping = mapping.map_input_indices(symbolc_shape)
+    else:
+        assert (
+            mapping.is_input_identity()
+        ), "non-identity input mapping is not supported yet"
+        index_mapping = mapping.map_output_indices(symbolc_shape)
+
+    index_mapping = tuple(subs_idxc(i) for i in index_mapping)
+
+    iters = mapping.iters
+
+    subs = [(sym, expr.start) for sym, expr in zip(iters.keys(), index.values())]
+
+    result_index = {key: m.subs(subs) for key, m in zip(symbolc_shape, index_mapping)}
+
+    # Iterate over elements_per_thread end check if every subsequent read have
+    # diff 1 in fastest changing dim and 0s in every other.
+    expected_diff = [0] * len(index_mapping)
+    expected_diff[-1] = 1
+
+    # Assume fastest changing dim increments in elements_per_thread between individual ops,
+    # This is needed for sympy simplifications.
+    subs[-1] = (subs[-1][0], (subs[-1][1] // elements_per_thread) * elements_per_thread)
```

Comment:
Could you add a comment explaining what sympy simplifications this enables?

Reply:
Expanded comments somewhat

Reply:
thanks